### ISSUE TITLE
Decision Trees

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#8833]
Decision Trees<br>
<br>
https://ift.tt/9d6WQJA<br>
<br>
<br>
<br>
<div>
<div>
<div>
<div>
<section>
<p>We just saw how a Decision Tree operates at a high-level: from the top down, it creates a series of sequential rules that split the data into well-separated regions for classification. But given the large number of potential options, how exactly does the algorithm determine where to partition the data? Before we learn how that works, we need to understand <span>Entropy</span>.<br>
<br>
Entropy measures the amount of information of some variable or event. We'll make use of it to identify regions consisting of a large number of similar (pure) or dissimilar (impure) elements.</p>
<p>Given a certain set of events that occur with probabilities , the total entropy can be written as the negative sum of weighted probabilities:</p>
<p>The quantity has a number of interesting properties:</p>
<div>
<h3>Entropy Properties</h3>
<ol>
<li>only if all but one of the are zero, this one having the value of 1. Thus the entropy vanishes only when there is no uncertainty in the outcome, meaning that the sample is completely unsurprising.</li>
<li>is maximum when all the are equal. This is the most uncertain, or 'impure', situation.</li>
<li>Any change towards the equalization of the probabilities increases .</li>
</ol>
</div>
<p>The entropy can be used to quantify the <span>impurity</span> of a collection of labeled data points: a node containing multiple classes is impure whereas a node including only one class is pure.</p>
<p>Above, you can compute the entropy of a collection of labeled data points belonging to two classes, which is typical for <span>binary classification</span> problems. Click on the <span>Add</span> and <span>Remove</span> buttons to modify the composition of the bubble.</p>
<p>Did you notice that pure samples have zero entropy whereas impure ones have larger entropy values? This is what entropy is doing for us: measuring how pure (or impure) a set of samples is. We'll use it in the algorithm to train Decision Trees by defining the Information Gain.</p>
</section>
<section>
<p>With the intuition gained with the above animation, we can now describe the logic to train Decision Trees. As the name implies, information gain measures an amount the information that we gain. It does so using entropy. <span>The idea is to subtract from the entropy of our data before the split the entropy of each possible partition thereafter</span>. We then select the split that yields the largest reduction in entropy, or equivalently, the largest increase in information.<br>
<br>
The core algorithm to calculate information gain is called <a href="https://en.wikipedia.org/wiki/ID3_algorithm">ID3</a>. It's a recursive procedure that starts from the root node of the tree and iterates top-down on all non-leaf branches in a greedy manner, calculating at each depth the difference in entropy:<br>
<br></p>
<p>To be specific, the algorithm's steps are as follows:</p>
<div>
<h3>ID3 Algorithm Steps</h3>
<ol>
<li>Calculate the entropy associated to every feature of the data set.</li>
<li>Partition the data set into subsets using different features and cutoff values. For each, compute the information gain as the difference in entropy before and after the split using the formula above. For the total entropy of all children nodes after the split, use the weighted average taking into account , i.e. how many of the samples end up on each child branch.</li>
<li>Identify the partition that leads to the maximum information gain. Create a decision node on that feature and split value.</li>
<li>When no further splits can be done on a subset, create a leaf node and label it with the most common class of the data points within it if doing classification or with the average value if doing regression.</li>
<li>Recurse on all subsets. Recursion stops if after a split all elements in a child node are of the same type. Additional stopping conditions may be imposed, such as requiring a minimum number of samples per leaf to continue splitting, or finishing when the trained tree has reached a given maximum depth.</li>
</ol>
</div>
<p>Of course, reading the steps of an algorithm isn't always the most intuitive thing. To make things easier to understand, let's revisit how information gain was used to determine the first <span>decision node</span> in our tree.</p>
<div>
<div>
<p>Recall our first decision node split on <span>Diameter &le; 0.45</span>. How did we choose this condition? It was the result of <span>maximizing information gain</span>.<br>
<br>
Each of the possible splits of the data on its two features (<span>Diameter</span> and <span>Height</span>) and cutoff values yields a different value of the information gain.<br>
<br>
The line chart displays the different split values for the <span>Diameter</span> feature. <span>Move the decision boundary yourself</span> to see how the data points in the top chart are assigned to the left or right children nodes accordingly. On the bottom you can see the corresponding entropy values of both children nodes as well as the total information gain.<br>
<br>
The ID3 algorithm will select the split point with the largest information gain, shown as the peak of the black line in the bottom chart of <span>0.574</span> at <span>Diameter = 0.45</span>.</p>
</div>
</div>
<p>Recall our first decision node split on <span>Diameter &le; 0.45</span>. How did we choose this condition? It was the result of <span>maximizing information gain</span>.<br>
<br>
Each of the possible splits of the data on its two features (<span>Diameter</span> and <span>Height</span>) and cutoff values yields a different value of the information gain.<br>
<br>
The visualization on the right allows to try different split values for the <span>Diameter</span> feature. <span>Move the decision boundary yourself</span> to see how the data points in the top chart are assigned to the left or right children nodes accordingly. On the bottom you can see the corresponding entropy values of both children nodes as well as the total information gain.<br>
<br>
The ID3 algorithm will select the split point with the largest information gain, shown as the peak of the black line in the bottom chart of <span>0.574</span> at <span>Diameter = 0.45</span>.</p>
<p>An alternative to the entropy for the construction of Decision Trees is the <a href="https://en.wikipedia.org/wiki/Decision_tree_learning#Gini_impurity">Gini impurity</a>. This quantity is also a measure of information and can be seen as a variation of Shannon's entropy. Decision trees trained using entropy or Gini impurity are comparable, and only in a few cases do results differ considerably. In the case of imbalanced data sets, entropy might be more prudent. Yet Gini might train faster as it does not make use of logarithms.</p>
</section>
<section>
<p>Let's recap what we've learned so far. First, we saw how a Decision Tree classifies data by repeatedly partitioning the feature space into regions according to some conditional series of rules. Second, we learned about entropy, a popular metric used to measure the purity (or lack thereof) of a given sample of data. Third, we learned how Decision Trees use entropy in information gain and the ID3 algorithm to determine the exact conditional series of rules to select. Taken together, the three sections detail the typical Decision Tree algorithm.<br>
<br>
To reinforce concepts, let's look at our Decision Tree from a slightly different perspective.<br>
<br>
The tree below maps exactly to the tree we showed in <span>How to Build a Decision Tree</span> section above. However, instead of showing the partitioned feature space alongside our trees structure, let's look at the partitioned data points and their corresponding entropy at each node itself:</p>
<br>
<p>From the top down, our sample of data points to classify shrinks as it gets partitioned to different <span>decision</span> and <span>leaf</span> nodes. In this manner, we could trace the full path taken by a training data point if we so desired. Note also that not every leaf node is pure: as discussed previously (and in the next section), we don't want the structure of our Decision Trees to be too deep, as such a model likely won't generalize well to unseen data.</p>
</section>
<section>
<p>Without question, Decision Trees have a lot of things going for them. They're simple models that are easy to interpret. They're fast to train and require minimal data preprocessing. And they hand outliers with ease. Yet they suffer from a major limitation, and that is their instability compared with other predictors. They can be <span>extremely sensitive to small perturbations in the data</span>: a minor change in the training examples can result in a drastic change in the structure of the Decision Tree.</p>
<p>Check for yourself how small random Gaussian perturbations on just 5% of the training examples create a set of completely different Decision Trees:</p>
<br>
<br>
<p>In their vanilla form, Decision Trees are unstable.<br>
<br>
If left unchecked, the ID3 algorithm to train Decision Trees will work endlessly to minimize entropy. It will continue splitting the data until all leaf nodes are completely pure - that is, consisting of only one class. Such a process may yield very deep and complex Decision Trees. In addition, we just saw that Decision Trees are subject to high variance when exposed to small perturbations of the training data.<br>
<br>
Both issues are undesirable, as they lead to predictors that fail to clearly distinguish between persistent and random patterns in the data, a problem known as <span>overfitting</span>. This is problematic because it means that our model won't perform well when exposed to new data.</p>
<p>There are ways to prevent excessive growth of Decision Trees by pruning them, for instance constraining their maximum depth, limiting the number of leaves that can be created, or setting a minimum size for the amount of items in each leaf and not allowing leaves with too few items in them.<br>
<br>
As for the issue of high variance? Well, unfortunately it's an intrinsic characteristic when training a single Decision Tree.</p>
</section>
<section>
<p>Perhaps ironically, one way to alleviate the instability induced by perturbations is to introduce an extra layer of randomness in the training process. In practice this can be achieved by creating <span>collections of Decision Trees</span> trained on slightly different versions of the data set, the combined predictions of which do not suffer so heavily from high variance. This approach opens the door to one of the most successful Machine Learning algorithms thus far: random forests.<br>
<br>
Stay tuned for our future article!</p>
</section>
<section>
<p>Thanks for reading! We hope that the article is insightful no matter where you are along your Machine Learning journey, and that you came away with a better understanding of the Decision Tree algorithm.<br>
<br>
To make things compact, we skipped over some relevant topics, such as using Decision Trees for regression, end-cut preference in tree models, and other tree-specific hyperparameters. Check out the resources listed below to learn more about those topics.<br>
<br>
To learn more about Machine Learning, check out our <a href="https://aws.amazon.com/machine-learning/mlu/">self-paced courses</a>, our <a href="https://www.youtube.com/channel/UC12LqyqTQYbXatYS9AA7Nuw">YouTube videos</a>, and the <a href="https://d2l.ai/">Dive into Deep Learning</a> textbook. If you have any comments or ideas related to <a href="https://mlu-explain.github.io/">MLU-Explain articles</a>, feel free to reach out directly to <a href="https://twitter.com/jdwlbr">Jared</a> or <a href="https://twitter.com/lusantala">Luc&iacute;a</a>. The code for this article is available <a href="https://github.com/aws-samples/aws-mlu-explain">here</a>.</p>
<p>A special thanks goes out to <span>Brent Werness</span> for valuable contributions to this article.</p>
<br>
<br>
<hr>
<br>
<br>
<p>This article is a product of the following resources + the awesome people who made (&amp; contributed to) them:</p>
<p><a href="https://onlinelibrary.wiley.com/doi/10.1002/j.1538-7305.1948.tb01338.x">A Mathematical Theory Of Communication</a><br>
(Claude E. Shannon, 1948).</p>
<p><a href="https://link.springer.com/article/10.1007/BF00116251">Induction of decision trees</a><br>
(John Ross Quinlan, 1986).</p>
<p><a href="https://www.dcc.fc.up.pt/~ltorgo/Papers/SECRTs.pdf">A Study on End-Cut Preference in Least Squares Regression Trees</a><br>
(Luis Torgo, 2001).</p>
<p><a href="https://link.springer.com/article/10.1007%2Fs10888-011-9188-x">The Origins Of The Gini Index: Extracts From Variabilit&agrave; e Mutabilit&agrave; (Corrado Gini, 1912)</a><br>
(Lidia Ceriani &amp; Paolo Verne, 2012).</p>
<p><a href="https://d3js.org/">D3.js</a><br>
(Mike Bostock &amp; Philippe Rivi&egrave;re)</p>
<p><a href="https://github.com/susielu/d3-annotation">d3-annotation</a><br>
(Susie Lu)</p>
<p><a href="https://katex.org/">KaTeX</a><br>
(Emily Eisenberg &amp; Sophie Alpert)</p>
<br>
<br></section>
</div>
</div>
</div>
</div>
<br>
<br>
<br>
<br>
<br>
<br>
via GitHub Pages | Websites for you and your projects, hosted directly from your GitHub repository. Just edit, push, and your changes are live.<br>
<br>
September 21, 2022 at 12:09AM
